pipelines/controller/src/main/java/org/openmrs/analytics/PipelineManager.java

```diff
@@ -315,56 +309,73 @@ synchronized void runIncrementalPipeline()
     if (!dataProperties.isCreateHiveResourceTables()) {
       return;
     }
 
-    DatabaseConfiguration dbConfig =
-        DatabaseConfiguration.createConfigFromFile(dataProperties.getThriftserverHiveConfig());
-
+    DatabaseConfiguration dbConfig;
+    try {
+      dbConfig =
+          DatabaseConfiguration.createConfigFromFile(dataProperties.getThriftserverHiveConfig());
+    } catch (IOException e) {
+      logger.error("Exception while reading thrift hive config.");
+      throw new RuntimeException(e);
+    }
     HiveTableManager hiveTableManager =
         new HiveTableManager(
             dbConfig.makeJdbsUrlFromConfig(),
             dbConfig.getDatabaseUser(),
             dbConfig.getDatabasePassword());
 
-    // Traverse through directories in parquet output directory to collect their names.
-    List<Path> paths;
-    try (Stream<Path> walk =
-        Files.walk(Paths.get(Constants.THRIFT_CONTAINER_PARQUET_PATH_PREFIX))) {
-      paths = walk.filter(Files::isDirectory).collect(Collectors.toList());
-    }
-    Map<String, List<String>> snapshots = new HashMap<>();
-    for (Path path : paths) {
-      String[] tokens = path.toString().split("/");
-      String resource = tokens[tokens.length - 1];
-      if (tokens.length > 0 && dataProperties.getResourceList().indexOf(resource) != -1) {
-        if (!snapshots.containsKey(resource)) {
-          snapshots.put(resource, new ArrayList<>());
-        }
-        snapshots.get(resource).add(path.toString());
+    String rootPrefix = dataProperties.getDwhRootPrefix();
+    Preconditions.checkState(rootPrefix != null && !rootPrefix.isEmpty());
 
-        // Get timestamp from directory name.
-        String[] snapshotsToken = tokens[tokens.length - 2].split(DataProperties.TIMESTAMP_PREFIX);
-        hiveTableManager.createResourceTable(resource, snapshotsToken[1], path.toString());
-      }
-    }
+    String prefix = getPrefix(rootPrefix);
+    String baseDir = getBaseDir(rootPrefix);
+    List<ResourceId> paths;
+    try {
+      paths =
+          getAllChildDirectories(baseDir).stream()
+              .filter(dir -> dir.getFilename().startsWith(prefix))
+              .collect(Collectors.toList());
+
+      Preconditions.checkState(paths != null, "Make sure DWH prefix is a valid path!");
 
-    // For each resource, create canonical table.
-    for (String resource : snapshots.keySet()) {
-      List<String> resourceSnapShots = snapshots.get(resource);
-      if (resourceSnapShots != null && !resourceSnapShots.isEmpty()) {
-        // Sort the directory names to make sure canonical table is created from the latest
-        // snapshot.
-        resourceSnapShots.sort(Collections.reverseOrder());
-        hiveTableManager.createResourceCanonicalTable(resource, resourceSnapShots.get(0));
+      int snapshotCount = 0;
+      for (ResourceId path : paths) {
+        if (!path.getFilename().startsWith(prefix + DataProperties.TIMESTAMP_PREFIX)) {
```

Comment:
nit: Consider moving `DataProperties.TIMESTAMP_PREFIX` into the `startsWith` filter that you have above (line 342) and remove this `if` block just to simplify the code (unless you think this is a likely issue that needs to be caught/warned about).

Reply:
done